src/storage/src/source/sql_server/replication.rs

```diff
@@ -100,38 +100,34 @@ pub(crate) fn render<G: Scope<Timestamp = Lsn>>(
                 .await?;
             let mut client = mz_sql_server_util::Client::connect(connection_config).await?;
 
-            let output_indexes: Vec<_> = outputs
-                .values()
-                .map(|v| usize::cast_from(v.partition_index))
-                .collect();
+            // The decoder is specific to the export, and each export pulls data from a specific capture instance.
+            let mut decoder_map: BTreeMap<_, _> = BTreeMap::new();
+            // Maps the 'capture instance' to the output index for only those outputs that this worker will snapshot
+            let mut export_ids_to_snapshot: BTreeMap<_, Vec<_>> = BTreeMap::new();
+            // Maps the 'capture instance' to the output index for all outputs of this worker
+            let mut capture_instances: BTreeMap<_, Vec<_>> = BTreeMap::new();
 
-            // Instances that have already made progress do not need to be snapshotted.
-            let needs_snapshot: BTreeSet<_> = outputs
-                .values()
-                .filter_map(|output| {
-                    if *output.resume_upper == [Lsn::minimum()] {
-                        Some(Arc::clone(&output.capture_instance))
-                    } else {
-                        None
-                    }
-                })
-                .collect();
+            for output in outputs.values() {
+                if decoder_map.insert(output.partition_index, Arc::clone(&output.decoder)).is_some() {
+                    panic!("Multiple decoders for output index {}", output.partition_index);
+                }
+                capture_instances.entry(Arc::clone(&output.capture_instance))
+                    .and_modify(|export_ids| export_ids.push(output.partition_index))
+                    .or_insert_with(|| vec![output.partition_index]);
```

Comment:
There is a cute API for this pattern

```suggestion
                capture_instances
                    .entry(Arc::clone(&output.capture_instance))
                    .or_default()
                    .push(output.partition_index);
```

`.or_default()` inserts the default value if it's not there and then returns a mutable reference to it so that you write your update logic once.